projects/hathor-wallet-headless/0005-early-send-tx-lock-release.md

```diff
@@ -0,0 +1,96 @@
+- Feature Name: early_send_tx_lock_release
+- Start Date: 2024-09-17
+- Author: Andre Carneiro <andreluizmrcarneiro@gmail.com>
+
+# Summary
+[summary]: #summary
+
+Release the send-tx lock after the transaction is prepared and before it is mined.
+
+# Motivation
+[motivation]: #motivation
+
+The only impediment to running 2 send transaction simultaneously is trying to choose the UTXOs and selecting the same ones.
+Meaning it is safe to release the send-tx lock after the transaction is chosen.
+This would allow sending more transactions from the same headless wallet instance.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+## Wallet-lib autorun
+
+The wallet facace has some methods that are meant to prepare and send transactions, these methods return a promise that resolves when the transaction is accepted by the network.
+This promise is the return of either `run` or `runFromMining` from the `SendTransaction` facade.
+
+The methods that send transactions are:
+
+- `sendManyOutputsTransaction`
+- `handleSendPreparedTransaction`
+- `consolidateUtxos`
+- `sendTransaction`
+- `createNewToken`
+- `mintTokens`
+- `meltTokens`
+- `delegateAuthority`
+- `destroyAuthority`
+- `createNFT`
+- `createNanoContractTransaction`
+- `createAndSendNanoContractTransaction`
+
+These methods (aside from `handleSendPreparedTransaction`) receive an `options` argument, we will add an `autorun` boolean that defaults to `true`.
+When `autorun` is `false` the method should return the `SendTransaction` instance instead.
+This will give the caller more control over the execution of the "send transaction" process.
```

Comment:
We already have an options parameter `startMiningTx` in these methods that defaults to `true` and I believe was added for this situation. I didn't see any use for this parameter so far but you can use it, since it already exists.

Apart from that, I'm not 100% comfortable with changing the return type of a method depending on a parameter value but I don't see any better way of doing this.

Reply:
`startMiningTx` is not being used at all and it is not available on all methods that we need.

Changing the return type is not ideal but the alternative would be duplicate all methods.
When I say duplicate I mean that we would create a method1 that returns the `SendTransaction` and a method2 that calls method1 and runs the `SendTransaction` returning the promise for basically each method of the list.
This duplication may be better, but having 1 method seems less complicated.

Reply:
I think I like the proposal but I feel it would break compatibility with all the wallets. Is that correct?

Reply:
Anything that changes the usual `return Promise<Transaction>` may break compatibility.
I don't think that the desktop or mobile use most of these methods, but we may require some refactoring.

Reply:
Changed the design to remove the autorun option and have a method that returns the `SendTransaction` instance, keeping backwards compatibility.